File: Axiom/Internal/Intermediate/ImBinaryOperation.cs
using System.IO;

namespace Axiom.Internal.Intermediate
{
    internal class ImBinaryOperation : ImExpression
    {
        internal enum BinaryOperator
        {
            Add, Subtract, Multiply, Divide, Modulo
        }

        public ImBinaryOperation(BinaryOperator oper, ImExpression left, ImExpression right)
        {
            Operator = oper;
            Left = left;
            Right = right;
        }

        public override void Dump(TextWriter tw, int ident)
        {
            Report.Dump(tw, ident, "BinaryOperation(" + Operator + ")");
            Left.Dump(tw, ident + 1);
            Right.Dump(tw, ident + 1);

        }

        public override ImExpressionSequence Linear()
        {
            var left = Left.Linear();
            var right = Right.Linear();
            var stmt = new ImSequence();

            stmt.Statements.AddAll(((ImSequence)left.Statement).Statements);
            stmt.Statements.AddAll(((ImSequence)right.Statement).Statements);

            return new ImExpressionSequence(stmt, new ImBinaryOperation(Operator, left.Expression, right.Expression));
        }

        public BinaryOperator Operator { get; private set; }
        public ImExpression Left { get; private set; }
        public ImExpression Right { get; private set; }

        internal static BinaryOperator OperatorFromToken(Token token)
        {
            switch (token) {
            case Token.Add:
                return BinaryOperator.Add;
            case Token.Subtract:
                return BinaryOperator.Subtract;
            case Token.Multiply:
                return BinaryOperator.Multiply;
            case Token.Divide:
                return BinaryOperator.Divide;
            case Token.Modulo:
                return BinaryOperator.Modulo;
            }

            Error.Report();

            return 0;
        }
    }
}

File: src/main/java/com/moybl/axiom/im/ImBinaryOperation.java
package com.moybl.axiom.im;

public class ImBinaryOperation implements ImExpression {

	public enum Operator {
		ADD,
		SUBTRACT,
		MULTIPLY,
		DIVIDE,
		MODULO
	}

	private Operator operator;
	private ImExpression left;
	private ImExpression right;

	public ImBinaryOperation(Operator operator, ImExpression left, ImExpression right) {
		this.operator = operator;
		this.left = left;
		this.right = right;
	}

	public Operator getOperator() {
		return operator;
	}

	public ImExpression getLeft() {
		return left;
	}

	public ImExpression getRight() {
		return right;
	}

	@Override
	public void accept(ImVisitor visitor) {
		visitor.visit(this);
	}

}
